README.md
# Eclipse Java Code Template Dynamic Variables Plug-in

Eclipse Plug-in to allow templates actions based on Java Bean fields present in a target class

See recent activity in the [Change Log](CHANGELOG.md)

# Installation

The Template Dynamic Variables plug-in is hosted via Eclipse update site at the address:

`https://raw.githubusercontent.com/StarChart-Labs/eclipse-dynamic-templates/latest-release/org.starchartlabs.eclipse.template.dynamic.site/`

# Use

This plug-in added two variables to the available variables for use in Java code templates. 

## enclosed_fields

The field `enclosed_fields` may be used to subsitute code per occurance of a field within the class being edited.

The variable is used in the form

`${id:enclosed_fields(template, separator)}`

The creator of the template replaces the following in the above:

- `id`
  - An identifier and default value (unique within the template) to use if there is a resolution error
- `template`
  - Code to subsitute per field within the enclosing class. Users may use the values:
    - `${type}` to substitute the field type
    - `${name}` to substitute the field name
- `separator`
  - Code to place between each occurance of the template generated. Users may use the value `${newline}` to substitute a newline into the separator. If code formatting is enabled for templates, its application may override the resulting newlines
  
## Example

The following template uses the `enclosed_fields` variable:

```
public ${enclosing_type}(${paramfields:enclosed_fields('${type} ${name}', ', ')}){
  ${assignfields:enclosed_fields('this.${name} = ${name};', '${newline}')}
}
```

If used within the class:

```
public class Example {

  private String field;
  
  private boolean boolField;
  
}
```

The template would result in the code:

```
public Example(String field, boolean boolField){
  this.field = field;
  this.boolField = boolField;
}
```

## enclosed_bean_fields

The field `enclosed_bean_fields` may be used to subsitute code per occurance of a field with a matching-named getter within the class being edited. A matching-name getter has no parameters, and is named `get(field)`, with the field's first letter capitalized. Boolean fields also match against `is(field)`.

The variable is used in the form

`${id:enclosed_bean_fields(template, separator)}`

The creator of the template replaces the following in the above:

- `id`
  - An identifier and default value (unique within the template) to use if there is a resolution error
- `template`
  - Code to subsitute per field/getter set within the enclosing class. Users may use the values:
    - `${type}` to substitute the field type
    - `${name}` to substitute the field name
    - `${getter}` to substitute the corresponding getter name of the field/getter pair
    - `${setter_name}` to substitute the corresponding setter name of of the field, if it exists (will not be subsituted if not present, and does not include parentheses or arguments)
- `separator`
  - Code to place between each occurance of the template generated. Users may use the value `${newline}` to substitute a newline into the separator. If code formatting is enabled for templates, its application may override the resulting newlines
  
## Example

The following template uses the `enclosed_bean_fields` variable:

```
public String toString(){
  return getClass().getSimpleName() + "{"
    + ${strfields:enclosed_bean_fields('"${name}=" + ${getter}', ' + "," ${newline}+ ')}
    "}";
}
```

If used within the class:

```
public class Example {

  private String field;
  
  private boolean boolField;
  
  private String noMatchedGetter;
  
  public String getField(){
    return field;
  }
  
  public boolean isBoolField(){
    return boolField;
  }
  
}
```

The template would result in the code:

```
public String toString(){
  return getClass().getSimpleName() + "{"
    + "field=" + getField() + ","
    + "boolField=" + isBoolField()
    "}";
}
```

# Contribution

Contributions are always welcome! 

* An Eclipse plug-in development environment is required to work with the projects - the last used Eclipse version was Oxygen. 
* The plug-ins are currently developed against Java 8
* All pull requests should be done against the master branch

See the [contribution documentation](./CONTRIBUTING.md) for detailed setup information and guidelines

## Licensing

The repository working set plug in is licensed under the Eclipse Public License 1.0, as specified and documented in the LICENSE.md file within each plug-in



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.3.0.0]
### Added
- Added "enclosed_fields" variable for line-per-field capabilities to Java code templates

## [0.2.0.0]
### Added

- Add ability to template insertion of setters (desprez)
- Add ability to template insertion of field type (desprez)

## [0.1.0.0]
### Added

- Added "enclosed_bean_fields" variable for line-per-field capabilities to Java code templates


org.starchartlabs.eclipse.template.dynamic/src/main/java/org/starchartlabs/eclipse/template/dynamic/EnclosingBeanFieldsResolver.java
/*
 * Copyright (c) Feb 28, 2018 StarChart Labs Authors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    romeara - initial API and implementation and/or initial documentation
 */
package org.starchartlabs.eclipse.template.dynamic;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.eclipse.jdt.core.IField;
import org.eclipse.jdt.core.IJavaElement;
import org.eclipse.jdt.core.IMethod;
import org.eclipse.jdt.core.IType;
import org.eclipse.jdt.core.JavaModelException;
import org.eclipse.jdt.internal.corext.template.java.CompilationUnitContext;
import org.eclipse.jface.text.templates.TemplateContext;
import org.eclipse.jface.text.templates.TemplateVariable;
import org.eclipse.jface.text.templates.TemplateVariableResolver;

/**
 * Fields resolver which allows defining a template to fill for every Java "bean" field
 *
 * <p>
 * In this context, a "bean" field is defined as a field which has a matching method called "get(Fieldname)". For
 * booleans, "is(Fieldname)" will also be considered a match
 *
 * <p>
 * References:
 * <ul>
 * <li>org.eclipse.jface.text.templates.TemplateVariable</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.CompilationUnitContextType</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.JavaContextType</li>
 * <li>https://stackoverflow.com/questions/350600/eclipse-custom-variable-for-java-code-templates</li>
 * <li>https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2Ftemplates%2FTemplateVariableResolver.html</li>
 * <li>http://help.eclipse.org/kepler/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Fconcepts%2Fconcept-template-variables.htm</li>
 * </ul>
 *
 * @author romeara
 */
@SuppressWarnings("restriction")
public class EnclosingBeanFieldsResolver extends TemplateVariableResolver {

    private static final String NAME_PLACEHOLDER = "\\$\\{name\\}";

    private static final String GETTER_PLACEHOLDER = "\\$\\{getter\\}";

    private static final Set<String> BOOLEAN_TYPE_SIGNATURES = Stream.of("Z", "QBoolean;").collect(Collectors.toSet());

    @Override
    public void resolve(TemplateVariable variable, TemplateContext context) {
        if (context instanceof CompilationUnitContext) {
            CompilationUnitContext jc = (CompilationUnitContext) context;

            String[] bindings = resolveAll(jc, variable.getVariableType().getParams());

            if (bindings != null) {
                variable.setValues(bindings);
                variable.setUnambiguous(true);
                variable.setResolved(true);
            } else {
                super.resolve(variable, context);
            }
        } else {
            super.resolve(variable, context);
        }
    }

    protected String[] resolveAll(CompilationUnitContext context, List<String> params) {
        String[] result = null;

        if (params.size() == 3) {
            String template = params.get(0);
            String separator = params.get(1);
            boolean newline = Boolean.valueOf(params.get(2));

            separator = (newline ? "\n" + separator : separator);

            List<String> lines = new ArrayList<>();

            for (Entry<String, String> entry : getBeanPairs(context).entrySet()) {
                lines.add(template
                        .replaceAll(NAME_PLACEHOLDER, entry.getKey())
                        .replaceAll(GETTER_PLACEHOLDER, entry.getValue()));
            }

            String value = lines.stream().collect(Collectors.joining(separator));
            result = new String[] { value };
        }

        return result;
    }

    protected Map<String, String> getBeanPairs(CompilationUnitContext context) {
        Map<String, String> result = new LinkedHashMap<String, String>();

        try {
            IType type = (IType) context.findEnclosingElement(IJavaElement.TYPE);
            Set<String> methodLookup = getCandidateMethods(type);

            for (IField field : type.getFields()) {
                String capitializedName = capitializeFirstLetter(field.getElementName());
                String getter = "get" + capitializedName;
                String isGetter = "is" + capitializedName;

                if (methodLookup.contains(getter)) {
                    result.put(field.getElementName(), getter + "()");
                } else if (isBooleanField(field) && methodLookup.contains(isGetter)) {
                    result.put(field.getElementName(), isGetter + "()");
                }
            }
        } catch (JavaModelException e) {
            throw new RuntimeException(e);
        }

        return result;
    }

    private Set<String> getCandidateMethods(IType type) throws JavaModelException {
        IMethod[] methods = type.getMethods();
        Set<String> methodLookup = new HashSet<>();

        for (IMethod method : methods) {
            if (method.getParameterNames().length == 0
                    && (method.getElementName().startsWith("get") || method.getElementName().startsWith("is"))) {
                methodLookup.add(method.getElementName());
            }
        }

        return methodLookup;
    }

    private boolean isBooleanField(IField field) throws JavaModelException {
        return BOOLEAN_TYPE_SIGNATURES.contains(field.getTypeSignature());
    }

    private String capitializeFirstLetter(String name) {
        String result = name;

        if(!name.isEmpty()) {
            String first = new String(name.substring(0, 1));
            result = name.replaceFirst(first, first.toUpperCase());
        }

        return result;
    }
}


org.starchartlabs.eclipse.template.dynamic/src/main/java/org/starchartlabs/eclipse/template/dynamic/resolver/EnclosedFieldsResolver.java
/*
 * Copyright (c) Feb 8, 2019 StarChart Labs Authors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    romeara - initial API and implementation and/or initial documentation
 */
package org.starchartlabs.eclipse.template.dynamic.resolver;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.stream.Collectors;

import org.eclipse.jdt.core.IField;
import org.eclipse.jdt.core.IJavaElement;
import org.eclipse.jdt.core.IType;
import org.eclipse.jdt.core.JavaModelException;
import org.eclipse.jdt.core.Signature;
import org.eclipse.jdt.internal.corext.template.java.CompilationUnitContext;
import org.eclipse.jface.text.templates.TemplateContext;
import org.eclipse.jface.text.templates.TemplateVariable;
import org.eclipse.jface.text.templates.TemplateVariableResolver;

/**
 * Fields resolver which allows defining a template to fill for every Java field
 *
 * <p>
 * The template variable resolved by this class is expected to be of the form:
 *
 * <pre>
 * ${id:enclosed_fields(template, separator)}
 * </pre>
 *
 * Where the user-provided values are:
 * <ul>
 * <li>id - identifier (unique within template) and default filler value in case of error</li>
 * <li>template - Line to substitute per bean-field. May use ${type} within to substitute field type, ${name} within to
 * substitute field name</li>
 * <li>separator - value, if any, to place between each occurrence of the substituted template. May use ${newline} to
 * substitute in a System.lineSeparator(). Resulting new-lines will be overridden by code formatter if it is enabled for
 * the template</li>
 * </ul>
 *
 * <p>
 * This class is intended to be extensible
 *
 * <p>
 * References:
 * <ul>
 * <li>org.eclipse.jface.text.templates.TemplateVariable</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.CompilationUnitContextType</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.JavaContextType</li>
 * <li>https://stackoverflow.com/questions/350600/eclipse-custom-variable-for-java-code-templates</li>
 * <li>https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2Ftemplates%2FTemplateVariableResolver.html</li>
 * <li>http://help.eclipse.org/kepler/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Fconcepts%2Fconcept-template-variables.htm</li>
 * </ul>
 *
 * @author romeara
 * @since 0.3.0
 */
@SuppressWarnings("restriction")
public class EnclosedFieldsResolver extends TemplateVariableResolver {

    protected static final String TYPE_PLACEHOLDER = "\\$\\{type\\}";

    protected static final String NAME_PLACEHOLDER = "\\$\\{name\\}";

    protected static final String NEWLINE_PLACEHOLDER = "\\$\\{newline\\}";

    @Override
    public void resolve(TemplateVariable variable, TemplateContext context) {
        if (context instanceof CompilationUnitContext) {
            CompilationUnitContext jc = (CompilationUnitContext) context;

            String[] bindings = resolveAll(jc, variable.getVariableType().getParams());

            // Store the result, and if resolved set unambiguous to true to avoid trying to get user input
            if (bindings != null) {
                variable.setValues(bindings);
                variable.setUnambiguous(true);
                variable.setResolved(true);
            } else {
                super.resolve(variable, context);
            }
        } else {
            super.resolve(variable, context);
        }
    }

    /**
     * Returns all possible bindings available in <code>context</code>, taking into account variable parameters
     *
     * @param context
     *            The Java compilation unit context in which to resolve the type
     * @param variableParameters
     * @return an array of possible bindings of this type in <code>context</code>, null if binding was unsuccessful
     */
    protected String[] resolveAll(CompilationUnitContext context, List<String> variableParameters) {
        String[] result = null;

        if (variableParameters.size() == 2) {
            String template = variableParameters.get(0);
            String separator = variableParameters.get(1).replaceAll(NEWLINE_PLACEHOLDER, System.lineSeparator());

            List<String> lines = new ArrayList<>();

            for (Entry<String, String> entry : getFieldTypes(context).entrySet()) {
                String processed = template
                        .replaceAll(TYPE_PLACEHOLDER, entry.getValue())
                        .replaceAll(NAME_PLACEHOLDER, entry.getKey());

                lines.add(processed);
            }

            String value = lines.stream().collect(Collectors.joining(separator));
            result = new String[] { value };
        }

        return result;
    }

    /**
     * Extracts field type information from a compilation unit
     *
     * @param context
     *            Information about the compilation unit the template is being inserted into
     * @return A mapping of any bean fields to their Java type
     */
    protected Map<String, String> getFieldTypes(CompilationUnitContext context) {
        Objects.requireNonNull(context);

        Map<String, String> result = new LinkedHashMap<>();

        try {
            IType type = (IType) context.findEnclosingElement(IJavaElement.TYPE);

            for (IField field : type.getFields()) {
                result.put(field.getElementName(), Signature.getSignatureSimpleName(field.getTypeSignature()));
            }
        } catch (JavaModelException e) {
            throw new RuntimeException(e);
        }

        return result;
    }

}


org.starchartlabs.eclipse.template.dynamic/src/main/java/org/starchartlabs/eclipse/template/dynamic/resolver/EnclosedBeanFieldsResolver.java
/*
 * Copyright (c) Feb 28, 2018 StarChart Labs Authors.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *    romeara - initial API and implementation and/or initial documentation
 *    desprez - add field type and setter.
 */
package org.starchartlabs.eclipse.template.dynamic.resolver;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.eclipse.jdt.core.IField;
import org.eclipse.jdt.core.IJavaElement;
import org.eclipse.jdt.core.IMethod;
import org.eclipse.jdt.core.IType;
import org.eclipse.jdt.core.JavaModelException;
import org.eclipse.jdt.core.Signature;
import org.eclipse.jdt.internal.corext.template.java.CompilationUnitContext;
import org.eclipse.jface.text.templates.TemplateContext;
import org.eclipse.jface.text.templates.TemplateVariable;
import org.eclipse.jface.text.templates.TemplateVariableResolver;

/**
 * Fields resolver which allows defining a template to fill for every Java "bean" field
 *
 * <p>
 * In this context, a "bean" field is defined as a field which has a matching method called "get(Fieldname)". For
 * booleans, "is(Fieldname)" will also be considered a match
 *
 * <p>
 * The template variable resolved by this class is expected to be of the form:
 *
 * <pre>
 * ${id:enclosed_bean_fields(template, separator)}
 * </pre>
 *
 * Where the user-provided values are:
 * <ul>
 * <li>id - identifier (unique within template) and default filler value in case of error</li>
 * <li>template - Line to substitute per bean-field. May use ${type} within to substitute field type, ${name} within to
 * substitute field name, ${getter} to substitute getter method (including parentheses), and ${setter_name} to
 * substitute the field's setter method name (if present)</li>
 * <li>separator - value, if any, to place between each occurrence of the substituted template. May use ${newline} to
 * substitute in a System.lineSeparator(). Resulting new-lines will be overridden by code formatter if it is enabled for
 * the template</li>
 * </ul>
 *
 * <p>
 * This class is intended to be extensible
 *
 * <p>
 * References:
 * <ul>
 * <li>org.eclipse.jface.text.templates.TemplateVariable</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.CompilationUnitContextType</li>
 * <li>org.eclipse.jdt.internal.corext.template.java.JavaContextType</li>
 * <li>https://stackoverflow.com/questions/350600/eclipse-custom-variable-for-java-code-templates</li>
 * <li>https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Ftext%2Ftemplates%2FTemplateVariableResolver.html</li>
 * <li>http://help.eclipse.org/kepler/index.jsp?topic=%2Forg.eclipse.jdt.doc.user%2Fconcepts%2Fconcept-template-variables.htm</li>
 * </ul>
 *
 * @author romeara, desprez
 * @since 0.1.0
 */
@SuppressWarnings("restriction")
public class EnclosedBeanFieldsResolver extends TemplateVariableResolver {

    protected static final String TYPE_PLACEHOLDER = "\\$\\{type\\}";

    protected static final String NAME_PLACEHOLDER = "\\$\\{name\\}";

    protected static final String GETTER_PLACEHOLDER = "\\$\\{getter\\}";

    protected static final String SETTER_PLACEHOLDER = "\\$\\{setter_name\\}";

    protected static final String NEWLINE_PLACEHOLDER = "\\$\\{newline\\}";

    private static final Set<String> BOOLEAN_TYPE_SIGNATURES = Stream.of("Z", "QBoolean;").collect(Collectors.toSet());

    @Override
    public void resolve(TemplateVariable variable, TemplateContext context) {
        if (context instanceof CompilationUnitContext) {
            CompilationUnitContext jc = (CompilationUnitContext) context;

            String[] bindings = resolveAll(jc, variable.getVariableType().getParams());

            // Store the result, and if resolved set unambiguous to true to avoid trying to get user input
            if (bindings != null) {
                variable.setValues(bindings);
                variable.setUnambiguous(true);
                variable.setResolved(true);
            } else {
                super.resolve(variable, context);
            }
        } else {
            super.resolve(variable, context);
        }
    }

    /**
     * Returns all possible bindings available in <code>context</code>, taking into account variable parameters
     *
     * @param context
     *            The Java compilation unit context in which to resolve the type
     * @param variableParameters
     * @return an array of possible bindings of this type in <code>context</code>, null if binding was unsuccessful
     */
    protected String[] resolveAll(CompilationUnitContext context, List<String> variableParameters) {
        String[] result = null;

        if (variableParameters.size() == 2) {
            String template = variableParameters.get(0);
            String separator = variableParameters.get(1).replaceAll(NEWLINE_PLACEHOLDER, System.lineSeparator());

            List<String> lines = new ArrayList<>();

            for (Entry<String, FieldInfo> entry : getBeanPairs(context).entrySet()) {
                FieldInfo fieldInfo = entry.getValue();

                String processed = fieldInfo.getSetter()
                        .map(setter -> template.replaceAll(SETTER_PLACEHOLDER, setter))
                        .orElse(template);

                processed = processed
                        .replaceAll(TYPE_PLACEHOLDER, fieldInfo.getType())
                        .replaceAll(NAME_PLACEHOLDER, entry.getKey())
                        .replaceAll(GETTER_PLACEHOLDER, fieldInfo.getGetter());

                lines.add(processed);
            }

            String value = lines.stream().collect(Collectors.joining(separator));
            result = new String[] { value };
        }

        return result;
    }

    /**
     * Finds all pairs of fields to methods which follow a pre-defined "bean" pattern. Pairs match this pattern if a
     * field has a corresponding method named "get(fieldname)". For booleans, "is(fieldname)" is also permissible. In
     * both cases the field name is expected to be capitalized
     *
     * @param context
     *            Information about the compilation unit the template is being inserted into
     * @return A mapping of any bean fields to the name of the method that matched
     */
    protected Map<String, FieldInfo> getBeanPairs(CompilationUnitContext context) {
        Objects.requireNonNull(context);

        Map<String, FieldInfo> result = new LinkedHashMap<String, FieldInfo>();

        try {
            IType type = (IType) context.findEnclosingElement(IJavaElement.TYPE);
            Set<String> methodLookup = getCandidateMethods(type);

            for (IField field : type.getFields()) {
                String capitalizedName = capitalizeFirstLetter(field.getElementName());

                String getter = "get" + capitalizedName;
                String isGetter = "is" + capitalizedName;

                FieldInfo fieldInfo = null;
                String setterMethod = "set" + capitalizedName;
                setterMethod = (methodLookup.contains(setterMethod) ? setterMethod : null);

                if (methodLookup.contains(getter)) {
                    fieldInfo = new FieldInfo(Signature.getSignatureSimpleName(field.getTypeSignature()), getter + "()",
                            setterMethod);
                } else if (isBooleanField(field) && methodLookup.contains(isGetter)) {
                    fieldInfo = new FieldInfo(Signature.getSignatureSimpleName(field.getTypeSignature()),
                            isGetter + "()", setterMethod);
                }

                if (fieldInfo != null) {
                    result.put(field.getElementName(), fieldInfo);
                }
            }
        } catch (JavaModelException e) {
            throw new RuntimeException(e);
        }

        return result;
    }

    /**
     * Finds and returns any methods within the provided type which match the expected pattern of a "getter" or
     * "setter". Matching "getter" methods start with either "get" or "is", and have no parameters. Matching "setter"
     * methods start with "set" and have one parameter
     *
     * @param type
     *            Eclipse JDT representation of the type the template is being inserted into
     * @return Set of method names matching "getter" criteria
     * @throws JavaModelException
     *             If there is an error reading Java model information from the type
     */
    private Set<String> getCandidateMethods(IType type) throws JavaModelException {
        Objects.requireNonNull(type);

        IMethod[] methods = type.getMethods();
        Set<String> methodLookup = new HashSet<>();

        for (IMethod method : methods) {
            if (method.getParameterNames().length == 0
                    && (method.getElementName().startsWith("get") || method.getElementName().startsWith("is"))) {
                methodLookup.add(method.getElementName());
            } else if (method.getParameterNames().length == 1 && method.getElementName().startsWith("set")) {
                methodLookup.add(method.getElementName());
            }
        }

        return methodLookup;
    }

    /**
     * Determines if a field represents a primitive or object boolean
     *
     * @param field
     *            The field to check
     * @return True if the field represents a form of Java boolean, false otherwise
     * @throws JavaModelException
     *             If there is an error reading Java model information from the field
     */
    private boolean isBooleanField(IField field) throws JavaModelException {
        Objects.requireNonNull(field);

        return BOOLEAN_TYPE_SIGNATURES.contains(field.getTypeSignature());
    }

    /**
     * Capitalizes the first letter in the provided string
     *
     * @param input
     *            The string to capitalize
     * @return The capitalized string
     */
    private String capitalizeFirstLetter(String input) {
        Objects.requireNonNull(input);

        String result = input;

        if (!input.isEmpty()) {
            String first = new String(input.substring(0, 1));
            result = input.replaceFirst(first, first.toUpperCase());
        }

        return result;
    }

}
